docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+
+3) Without making any assumptions on how & when to trigger `MatrixEvaluationStrategy` is not of any use as the grid splits are arbitrary and don't correspond to the times when triggers occur and models are transitioned.
+
+4) `OfflineEvaluationStrategy` is not flexible enough to handle arbitrary intervals. Even if we would run after the pipeline where we know the exact points in time when triggers occurred, we could still not specify the intervals as `OfflineEvaluationStrategy` only supports 1-sided intervals. It currently assumes being given `first_timestamp` and `last_timestamp` representing the first and last sample timestamp for the data batch that lead to the trigger. These do not correspond to the start and end time of the model usage.
+
+>>> The ideal solution would therefore evaluate models on their entire usage period with all samples in that period.
+
+```mermaid
+gantt
+    title A Gantt Diagram
+    dateFormat YYYY-MM-DD
+    axisFormat %Y
+    section models
+        Trigger1 :milestone, 1930, 0d
+        model 1          :a1, 1930, 10y
+        Trigger2 :milestone, :after a1
+        model 2          :a2, after a1, 5y
+        Trigger3 :milestone, :after a2
+        model 3          :a3, after a2, 5y
+        Trigger4 :milestone, :after a3
+        model 4          :a4, after a3, 25y
+        Trigger5 :milestone, :after a4
+        model 5          :a5, after a4, 25y
+        Trigger6 :milestone, :after a5
+        model 6          :a6, after a5, 5y
+        Trigger7 :milestone, :after a6
+        model 7          :a7, after a6, 5y
+    section evaluations
+        Evaluation1 :milestone, 1940, 0d
+        Evaluation2 :milestone, 1945, 0d
+        Evaluation3 :milestone, 1950, 0d
+        Evaluation4 :milestone, 1975, 0d
+        Evaluation5 :milestone, 2000, 0d
+        Evaluation6 :milestone, 2005, 0d
+        Evaluation7 :milestone, 2010, 0d
+    section evaluation data
+        data1 :e1, 1930, 10y
+        data2 :e2, after e1, 5y
+        data3 :e3, after e2, 5y
+        data4 :e4, after e3, 25y
+        data5 :e5, after e4, 25y
+        data6 :e6, after e5, 5y
+        data7 :e7, after e6, 5y
+```
+
+##### Proposed solution
+
+We add a new evaluation strategy with an according pipeline stage that runs evaluation after the rest of the core pipeline finished. By doing so we have access to all trigger times and can evaluate the models in the correct time ranges. After running evaluations on every "triggering-interval" (time between two triggers), we can calculate the pipeline performance as the weighted average of the model performances accounting for the number of samples in the time range.
+
+> Hypothesis: There's no real need to run evaluations directly after the training. We can always do that after the core pipeline finished. If required we can also make use of information like `first_timestamp` and `last_timestamp` from the logs for a specific trigger.
+
```

Comment:
Yes this is correct. I placed it there to have *early feedback*, i.e. without waiting for the whole pipeline to finish, I can already access some of the early model's evaluation results.

Reply:
Do you use the `early feedback` somewhere during the pipeline run?

Reply:
I just retrieve the pipeline log before the pipeline finishes. Then I can look into the data to ensure no bug is going on.

Reply:
yes there is a tradeoff here. In some sense it would be amazing to overlap more things in Modyn's execution model: We could overlap training a new model, while evaluation is still going on (assuming enough GPU resources etc), we could run evaluations in parallel given multiple GPU resources and whatnot. Right now Modyn's execution model is pretty sequential, but I agree nevertheless early evaluation is helpful to investigate the logs early.